decentraland/explorer
# feat: Global voice chat Experimental Feature

This PR adds functionality for a global voice chat implementation that uses the existing P2P network to send audio recorded in the browser, and then encoded using an Opus Codec. On the receiving end, each audio stream is decoded and played with spatial audio parameters for an immersive experience.

### To test this PR:
- Ensure you have the `VOICE_CHAT_ENABLED` parameter added in the URL. For example: https://play.decentraland.zone/branch/feat/global-voice-chat-poc/index.html?VOICE_CHAT_ENABLED&ENV=org
- Press T to talk. Voice chat is implemented as a "push to talk" feature so you'll need to keep T pressed for as long as you are talking.

### Review hints
The PR is long. Sorry about that :sweat_smile:. Some files like those related to libopusjs and the resampler are imported from external sources. They were no NPM packages. But you can think of them as external dependencies.

### What's Next (For future PRs)
- Usability and UI need to be reworked. We may want to add a recording toggle feature to work alongside the push to talk feature. 
- Mute support and voice chat volume control. Currently you can mute a person by blocking them
- More audio settings like device selection and all that
- Voice may not be as clear as it needs if network is not relatively stable or there is congestion. We may have to implement QoS protocol in the peer network
- We may want to implement a dynamic buffer that adjust to network jitter
- Currently, packets that arrive late (after a newer packet) are discarded. We may want to process them anyway and take advantage of the buffer to reorder the packages to improve reliability
- Spatial parameters need to be tuned. Currently, you can hear people that are quite far. We may want more attenuation or an ad-hoc mechanism to ignore audio from a certain distance. This distance may be dynamic, and depend on your friend list or other variables.
